zcash_client_sqlite/src/wallet.rs

```diff
@@ -656,47 +751,99 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
                     addr_str,
                 ))),
             })
-            .map(|addr| (addr, DiversifierIndex::from(di_be)))
+            .map(|addr| (addr, diversifier_index))
     })
     .transpose()
 }
 
-/// Adds the given address and diversifier index to the addresses table.
+/// Adds the given external address and diversifier index to the addresses table.
 ///
-/// Returns the database row for the newly-inserted address.
-pub(crate) fn insert_address<P: consensus::Parameters>(
+/// Returns the primary key identifier for the newly-inserted address.
+pub(crate) fn upsert_address<P: consensus::Parameters>(
     conn: &rusqlite::Connection,
     params: &P,
     account_id: AccountRef,
     diversifier_index: DiversifierIndex,
     address: &UnifiedAddress,
-) -> Result<(), SqliteClientError> {
+    exposed_at_height: Option<BlockHeight>,
+) -> Result<AddressRef, SqliteClientError> {
     let mut stmt = conn.prepare_cached(
         "INSERT INTO addresses (
             account_id,
             diversifier_index_be,
+            key_scope,
             address,
-            cached_transparent_receiver_address
+            transparent_child_index,
+            cached_transparent_receiver_address,
+            exposed_at_height
         )
         VALUES (
             :account_id,
             :diversifier_index_be,
+            :key_scope,
             :address,
-            :cached_transparent_receiver_address
-        )",
+            :transparent_child_index,
+            :cached_transparent_receiver_address,
+            :exposed_at_height
+        )
+        ON CONFLICT (account_id, diversifier_index_be, key_scope) DO UPDATE
+        SET exposed_at_height = :exposed_at_height
```

Comment:
This should be the minimum of the existing and new heights, with `NULLS` greatest (i.e. overwrite a `NULL` in either direction, and take the minimum if both are non-null).